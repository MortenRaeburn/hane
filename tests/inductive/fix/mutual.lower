Inductive tree (Set) : Set :=
    | Node : forall['0] (forall[(forest) ('1)] ((tree) ('2)))
    with forest (Set) : Set :=
    | Nil : (forest) ('0)
    | Cons : forall[(tree) ('0)] (forall[(forest) ('1)] ((forest) ('2))).
Definition map : forall[Set] (forall[Set] (forall[forall['1] ('1)] (forall[(tree) ('2)] ((tree) ('2))))) := fix[Set, Set, forall['1] ('1), (tree) ('2)]{struct 3} : (tree) ('2) := match '0 in tree return (tree) ('4) with (((Node) ('5)) (('4) ('1))) ((((('7) ('6)) ('5)) ('4)) ('0)) end with[Set, Set, forall['1] ('1), (forest) ('2)]{struct 3} : (forest) ('2) := match '0 in forest return (forest) ('4) with (Nil) ('3) | (((Cons) ('5)) ((((('8) ('6)) ('5)) ('4)) ('1))) ((((('7) ('6)) ('5)) ('4)) ('0)) end for 0.
