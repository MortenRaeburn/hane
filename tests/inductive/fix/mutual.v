Inductive tree (A : Set) : Set :=
    | Node : forall (x : A) (ts : forest A), tree A
    with forest (A : Set) :=
    | Nil : forest A
    | Cons : forall (t : tree A) (ts : forest A), forest A.

Definition map : forall (A : Set) (B : Set) (f : forall x : A, B) (t : tree A), tree B :=
    fix map_tree (A : Set) (B : Set) (f : forall x : A, B) (t : tree A) : tree B :=
        match t as _ in tree _ return tree B with
        | Node _ x ts => Node B (f x) (map_forest A B f ts)
        end
    with map_forest (A : Set) (B : Set) (f : forall x : A, B) (ts : forest A) : forest B :=
        match ts as _ in forest _ return forest B with
        | Nil _ => Nil B
        | Cons _ t ts => Cons B (map_tree A B f t) (map_forest A B f ts)
        end
    for map_tree.
