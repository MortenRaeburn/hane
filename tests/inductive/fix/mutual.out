Inductive tree (Set) : Set :=
    | Node : forall['0] (forall[(forest) ('1)] ((tree) ('2)))
    with forest (Set) : Set :=
    | Nil : (forest) ('0)
    | Cons : forall[(tree) ('0)] (forall[(forest) ('1)] ((forest) ('2))).
Definition map : forall[Set] (forall[Set] (forall[forall['1] ('1)] (forall[(tree) ('2)] ((tree) ('2))))) := fix[Set, Set, forall['1] ('1), (tree) ('2)] : (tree) ('2) := match '2 in tree return (tree) ('6) with (((Node) ('7)) (('6) ('1))) ((((('3) ('8)) ('7)) ('6)) ('0)) end with[Set, Set, forall['1] ('1), (forest) ('2)] : (forest) ('2) := match '2 in forest return (forest) ('6) with (Nil) ('5) | (((Cons) ('7)) ((((('4) ('8)) ('7)) ('6)) ('1))) ((((('3) ('8)) ('7)) ('6)) ('0)) end for 0.
